clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -84,56 +119,84 @@ public void setAssignorSelection(AssignorSelection assignorSelection) {
         this.assignorSelection = assignorSelection;
     }
 
+    /**
+     * Update the member state, setting it to the nextState only if it is a valid transition.
+     *
+     * @throws IllegalStateException If transitioning from the member {@link #state} to the
+     *                               nextState is not allowed as defined in {@link MemberState}.
+     */
     private void transitionTo(MemberState nextState) {
         if (!this.state.equals(nextState) && !nextState.getPreviousValidStates().contains(state)) {
-            // TODO: handle invalid state transition
-            throw new RuntimeException(String.format("Invalid state transition from %s to %s",
+            throw new IllegalStateException(String.format("Invalid state transition from %s to %s",
                     state, nextState));
         }
         this.state = nextState;
```

Comment:
Might I suggest some `log.trace()`-level logging when this state transition occurs?